src/midi/midi_fluidsynth.cpp

```diff
@@ -0,0 +1,245 @@
+
+	fluid_settings_setint(settings.get(), "synth.chorus.active",
+	                      section->Get_bool("fluid.chorus"));
+	fluid_settings_setint(settings.get(), "synth.chorus.nr",
+	                      section->Get_int("fluid.chorus.number"));
+	fluid_settings_setnum(settings.get(), "", get_double("fluid.chorus.level"));
+	fluid_settings_setnum(settings.get(), "synth.chorus.speed",
+	                      get_double("fluid.chorus.speed"));
+	fluid_settings_setnum(settings.get(), "synth.chorus.depth",
+	                      get_double("fluid.chorus.depth"));
+
+	fsynth_ptr_t fluid_synth(new_fluid_synth(settings.get()), delete_fluid_synth);
+	if (!fluid_synth) {
+		// log_cb(RETRO_LOG_WARN, "[dosbox] Error creating fluidsynth synthesiser\n");
+		return false;
+	}
+
+	std::string soundfont = section->Get_string("fluid.soundfont");
+	if (!soundfont.empty() && fluid_synth_sfcount(fluid_synth.get()) == 0) {
+		fluid_synth_sfload(fluid_synth.get(), soundfont.data(), true);
+	}
+
+	mixer_channel_ptr_t mixer_channel(
+	        MIXER_AddChannel(mixerCallback,
+	                         section->Get_int("fluid.samplerate"), "FSYNTH"),
+	        MIXER_DelChannel);
+	mixer_channel->Enable(true);
+
+	fluid_settings = std::move(settings);
+	synth = std::move(fluid_synth);
+	channel = std::move(mixer_channel);
+	is_open = true;
+	return true;
+}
+
+void MidiHandlerFluidsynth::Close()
+{
+	if (!is_open)
+		return;
+
+	channel->Enable(false);
+	channel = nullptr;
+	synth = nullptr;
+	fluid_settings = nullptr;
+	is_open = false;
+}
+
+void MidiHandlerFluidsynth::PlayMsg(const uint8_t *msg)
+{
+	const int chanID = msg[0] & 0b1111;
+
+	switch (msg[0] & 0b1111'0000) {
+	case 0b1000'0000:
+		fluid_synth_noteoff(synth.get(), chanID, msg[1]);
+		break;
+	case 0b1001'0000:
+		fluid_synth_noteon(synth.get(), chanID, msg[1], msg[2]);
+		break;
+	case 0b1010'0000:
+		fluid_synth_key_pressure(synth.get(), chanID, msg[1], msg[2]);
+		break;
+	case 0b1011'0000:
+		fluid_synth_cc(synth.get(), chanID, msg[1], msg[2]);
+		break;
+	case 0b1100'0000:
+		fluid_synth_program_change(synth.get(), chanID, msg[1]);
+		break;
+	case 0b1101'0000:
+		fluid_synth_channel_pressure(synth.get(), chanID, msg[1]);
+		break;
+	case 0b1110'0000:
+		fluid_synth_pitch_bend(synth.get(), chanID, msg[1] + (msg[2] << 7));
+		break;
+	default: {
+		uint64_t tmp;
+		memcpy(&tmp, msg, sizeof(tmp));
+		// log_cb(RETRO_LOG_WARN, "[dosbox] fluidsynth: unknown MIDI command: %08lx", tmp);
+		break;
+	}
+	}
+}
+
+void MidiHandlerFluidsynth::PlaySysex(uint8_t *sysex, size_t len)
+{
+	const char *data = reinterpret_cast<const char *>(sysex);
+	fluid_synth_sysex(synth.get(), data, len, nullptr, nullptr, nullptr, false);
+}
+
+// TODO: originally, buffer MixTemp was being used here to receive int16_t data
+// from fluid and pass it ot the mixer... but this buffer is uint8_t, thus
+// introduces alignment problem; replaced with new local buffer for now.
+int16_t data[MIXER_BUFSIZE];
+
```

Comment:
The size defined by MIXER_BUFSIZE is way too big (appears to be defined by someone who wasn't familiar with the 1ms timing of dosbox); suggest 96 * 2 (just incase someone really does set the rate to 96 KHz.. and then in the function below, `assert(len <= 96 * 2)`.. you'll find that it's never runs over :-)

I've done the same thing in the GUS mixer branch (where I've done away with `gusrate`, and therefore the higher sampling rate is a function of the number of active voices at 44.1 KHz, so I've set it's static buffer to 48 * 2 samples.. and it's never overrun in ~2 months of testing).